s3/s3_file.go

```diff
@@ -0,0 +1,266 @@
+	}
+	return names, err
+}
+
+// Stat returns the FileInfo structure describing file.
+// If there is an error, it will be of type *PathError.
+func (f *S3File) Stat() (os.FileInfo, error) {
+	return f.s3Fs.Stat(f.Name())
+}
+
+// Sync is a noop.
+func (f *S3File) Sync() error {
+	return nil
+}
+
+// Truncate changes the size of the file.
+// It does not change the I/O offset.
+// If there is an error, it will be of type *PathError.
+func (f *S3File) Truncate(size int64) error {
+	panic("implement Truncate")
+	return nil
+}
+
+// WriteString is like Write, but writes the contents of string s rather than
+// a slice of bytes.
+func (f *S3File) WriteString(s string) (int, error) {
+	return f.Write([]byte(s))
+}
+
+// Close closes the File, rendering it unusable for I/O.
+// It returns an error, if any.
+func (f *S3File) Close() error {
+	f.closed = true
+	return nil
+}
+
+// Read reads up to len(b) bytes from the File.
+// It returns the number of bytes read and an error, if any.
+// EOF is signaled by a zero count with err set to io.EOF.
+func (f *S3File) Read(p []byte) (int, error) {
+	if f.closed {
+		// mimic os.File's read after close behavior
+		panic("read after close")
+	}
+	if f.offset != 0 {
+		panic("TODO: non-offset == 0 read")
+	}
+	if len(p) == 0 {
+		return 0, nil
+	}
+	output, err := f.s3API.GetObject(&s3.GetObjectInput{
+		Bucket: aws.String(f.bucket),
+		Key:    aws.String(f.name),
+	})
+	if err != nil {
+		return 0, err
+	}
+	defer output.Body.Close()
+	n, err := output.Body.Read(p)
+	f.offset += n
+	return n, err
+}
+
+// ReadAt reads len(p) bytes from the file starting at byte offset off.
+// It returns the number of bytes read and the error, if any.
+// ReadAt always returns a non-nil error when n < len(b).
+// At end of file, that error is io.EOF.
+func (f *S3File) ReadAt(p []byte, off int64) (n int, err error) {
+	_, err = f.Seek(off, 0)
+	if err != nil {
+		return
+	}
+	n, err = f.Read(p)
+	return
+}
+
+// Seek sets the offset for the next Read or Write on file to offset, interpreted
+// according to whence: 0 means relative to the origin of the file, 1 means
+// relative to the current offset, and 2 means relative to the end.
+// It returns the new offset and an error, if any.
+// The behavior of Seek on a file opened with O_APPEND is not specified.
+func (f *S3File) Seek(offset int64, whence int) (int64, error) {
+	switch whence {
+	case 0:
+		f.offset = int(offset)
+	case 1:
+		f.offset += int(offset)
+	case 2:
+		// can probably do this if we had GetObjectOutput (ContentLength)
+		panic("TODO: whence == 2 seek")
+	}
+	return int64(f.offset), nil
+}
+
+// Write writes len(b) bytes to the File.
+// It returns the number of bytes written and an error, if any.
+// Write returns a non-nil error when n != len(b).
+func (f *S3File) Write(p []byte) (int, error) {
+	if f.closed {
+		// mimic os.File's write after close behavior
+		panic("write after close")
+	}
+	if f.offset != 0 {
+		panic("TODO: non-offset == 0 write")
```

Comment:
This seems like it should be fixed, otherwise Write in existing files is not useful?